### ISSUE TITLE
VictoriaLogs web UI: stuck browser

### ISSUE THREAD

**Issue by yuriydzobak** [VictoriaMetrics/VictoriaMetrics#7185]
### Describe the bug

Hi,
The VictoriaLogs UI can stuck or crash browser, when you have a log entry that has message with size more 200kb  in `_msg` field
When i'm using curl or vlogscli it's working fast and show data less then 1 second
```
;> _tume:2d pod_namespace:monitoring container_name:lf | limit 20;
executing [_tume:2d pod_namespace:monitoring container_name:lf | limit 20]...; duration: 0.650s
;> _tume:2d pod_namespace:monitoring container_name:lf | limit 20;
executing [_tume:2d pod_namespace:monitoring container_name:lf | limit 20]...; duration: 0.202s
;> _tume:2d pod_namespace:monitoring container_name:lf | limit 20;
executing [_tume:2d pod_namespace:monitoring container_name:lf | limit 20]...; duration: 0.193s
;> _tume:2d pod_namespace:monitoring container_name:lf | limit 20;
executing [_tume:2d pod_namespace:monitoring container_name:lf | limit 20]...; duration: 0.194s
;> _tume:2d pod_namespace:monitoring container_name:lf | limit 20;
executing [_tume:2d pod_namespace:monitoring container_name:lf | limit 20]...; duration: 0.192s
```

but if i do the same in VMUI it spends around 10-15 seconds or even browser stuck

### To Reproduce

1. Run VictoriaLogs go to https://github.com/VictoriaMetrics/VictoriaMetrics/tree/master/deployment/docker/victorialogs/vector/loki
2. change inputs in `vector.yaml` from
```
sinks:
  loki:
    type: loki
    inputs:
      - demo
```      
```
    inputs:
      - docker
```      
2. docker compose up -d
3. Run container that generates log message with around 200kb 
4. Create file `generate_logs.py`
```
import json
import random
import string
from datetime import datetime
import time

# Define constants
length_per_log = 204000  # Length of each log entry in bytes (200 KB)
max_line_size = 262144  # Max line size in bytes (256 KB)

# Ensure length_per_log does not exceed max_line_size
if length_per_log > max_line_size:
    length_per_log = max_line_size

# Calculate the maximum number of words based on length_per_log
max_words = (length_per_log // 2) - 1  # Each word is 2 characters long, subtract 1 for the last space

try:
    while True:
        # Generate a list of random words (2 characters each)
        log_message = ' '.join(''.join(random.choices(string.ascii_letters + string.digits, k=2)) for _ in range(max_words))

        # Create a log entry
        log_entry = {
            "timestamp": datetime.now().isoformat(),
            "container_name": f"container_{random.randint(0, 99)}",
            "host": f"host_{random.randint(0, 9)}",
            "source_type": f"source_{random.randint(0, 4)}",
            "msg": log_message
        }

        # Print log entry to stdout
        print(json.dumps(log_entry))

        # Sleep for a short duration (e.g., 0.1 seconds) to prevent high CPU usage
        time.sleep(0.1)

except KeyboardInterrupt:
    print("\nLog generation stopped.")
```
5. Create `dockerfile`
```
# Use a lightweight Python image
FROM python:3.9-slim

# Set the working directory
WORKDIR /app

# Copy the log generation script into the container
COPY generate_logs.py .

# Install any necessary dependencies (if any)
# For this example, there are no additional dependencies

# Run the log generation script
CMD ["python", "generate_logs.py"]
```
6. Build - `docker build -t log-generator .`
7. run - `docker run -d --name log-generator-container log-generator`
8. wait 2-5 minutes
8. open VMUI on `http://localhost:9428/select/vmui`
9. try to show logs for example 100-500 entries, i can't reproduce like in my case with 20 entries the browser stuck, maybe a log ingesting load or network speed =( 

https://youtu.be/iK88ngD2IPM
In UI the 100 entriesis loading a couple second, if i do it via curl it needs a small amount of time
```
curl http://localhost:9428/select/logsql/query -d 'query=_time:5m | limit 100  0.01s user 0.10s system 9% cpu 1.127 total
```

Another issue, As you can see in the end of video, if I set a huge limit, the UI hangs. I can't change the limit to a smaller value because the URL doesn't have this parameter. If I use a limit of 1k or more, the browser always crashes. The only solution is to clear the cache and cookies.

### Version

teted on 0.28.0 - 0.33.0

### Additional information

grafana plugin shows random such like entries https://github.com/VictoriaMetrics/victorialogs-datasource/issues/80

**Comment by hagen1778** [comment]
Looks like it is time to add complexity limiting cmd-line flags:
* `search.maxEntries`, int, limits the number of returned log entries
* `search.maxResponseSize`, int, limits the number of bytes in the returned response

wdyt @valyala @Loori-R ?

**Comment by Loori-R** [comment]
I will add the `limit` parameter to the URL.

As for long loading times, the following features should help avoid this:
- #7097
- #7117

Regarding the browser crash: I will investigate the issue and may optimize the response handling. However, we cannot fully avoid this since, with large data volumes, the browser may struggle to parse the response, and this is beyond our control.

**Comment by Loori-R** [comment]
@yuriydzobak FYI: 
You can specify the limit in the URL. If the limit is not specified in the URL, it defaults to 50. If the limit was changed previously, it will be saved in local storage and will be used from there. If the limit is specified in the URL, that limit will be applied.

For example:
http://localhost:3000/#/?query=*&g0.range_input=5m&g0.end_input=2024-10-08T14%3A50%3A16&g0.relative_time=last_5_minutes&limit=10

**Comment by yuriydzobak** [comment]
Hi,
First of all, thank you for fixing and implementing a safe mechanism like this one. https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7097

I'm using 0.37.0, It seems that this doesn't help when the logs are large, such as 200 KB, especially if the JSON inside another JSON(i configured wrong first time log collector). This might be the issue. For example, if you run a query for an app with 200 KB logs, the UI becomes unresponsive while loading the logs and may crash.
Example bellows, if needs some help or you can't reproduce it, i can try to create some simular logs via script 

https://github.com/user-attachments/assets/c294dec0-c9a8-443d-a8ab-16914304da63

**Comment by valyala** [comment]
> Looks like it is time to add complexity limiting cmd-line flags:

> `-search.maxEntries` - limits the number of returned log entries
> `-search.maxResponseSize` - limits the number of bytes in the returned response

> wdyt @valyala @Loori-R ?

I don't think these limits are needed, since they will complicate configuration of VictoriaLogs without visible benefits. The [`/select/logsql/query` HTTP endpoint](https://docs.victoriametrics.com/victorialogs/querying/#querying-logs) is designed in the way, which allows returning arbitrary large response using JSON line streaming. E.g. it returns every log entry as a single JSON line, and it can return arbitrary number of log entries (for example, if you need analyzing the returned log entries with some external application). As @Loori-R already said, the `/select/logsql/query` endpoint supports `limit` query arg (it is documented [here](https://docs.victoriametrics.com/victorialogs/querying/#querying-logs)), which allows limiting the number of returned JSON lines. This query arg is already passed by [VictoriaLogs web UI](https://docs.victoriametrics.com/victorialogs/querying/#web-ui) to `/select/logsql/query` endpoint. So it limits the number of log entries, which can be returned to web UI. Every log entry size is limited by on-disk data block size where log entries are stored. Currently it is set to 2MiB: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/f62502a9432dd8eb2b32ad9f0fc248633edc8abf/lib/logstorage/consts.go#L18-L21 . This effectively limits the maximum response size to `2MiB` * `limit`. By default web UI uses `limit=50`, so it may receive response from VictoriaLogs with the maximum size `2MiB * 50 = 100MiB`. IMHO, this isn't big number for processing by modern web browsers. This may lead to slowdown or OOM crash of web browser only if web UI contains some very inefficient code for processing responses of such sizes. This code must be detected and fixed.

**Comment by hagen1778** [comment]
@yuriydzobak @valyala a lot of UI improvements were added in https://github.com/VictoriaMetrics/VictoriaMetrics/commit/10c42668a1455a0adaa1172ea12179d975fbb7cb. Would you mind checking if that helps to resolve your issues? These changes will be included into the next release. But meanwhile, you can try them by building vlogs + ui from sources (requires docker):
```
make vmui-logs-update
make victoria-logs
```

**Comment by yuriydzobak** [comment]
> @yuriydzobak @valyala a lot of UI improvements were added in [10c4266](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/10c42668a1455a0adaa1172ea12179d975fbb7cb). Would you mind checking if that helps to resolve your issues? These changes will be included into the next release. But meanwhile, you can try them by building vlogs + ui from sources (requires docker):
> 
> ```
> make vmui-logs-update
> make victoria-logs
> ```

I'll try tomorrow and provide you with feedback

**Comment by yuriydzobak** [comment]
Hi,
I've tested it. The browser hasn't crashed in my case anymore, and UI works much better in first impression
Thanks you for improvement !
https://github.com/user-attachments/assets/eca2bb99-bae4-441d-b994-64ae37a08ca4

**Comment by valyala** [comment]
All the improvements from the commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/10c42668a1455a0adaa1172ea12179d975fbb7cb have been included in [v1.1.0-victorialogs](https://docs.victoriametrics.com/victorialogs/changelog/). Closing the issue as fixed then.
